symengine/logic.h

```diff
@@ -168,6 +168,97 @@ class Xor : public Boolean
     const vec_boolean &get_container() const;
 };
 
+class Relational : public TwoArgBasic<Boolean>
+{
+public:
+    //! Constructor
+    Relational(const RCP<const Basic> &lhs, const RCP<const Basic> &rhs);
+    //! \return the hash
+    hash_t __hash__() const;
+    /*! Equality comparator
+     * \param o - Object to be compared with
+     * \return whether the 2 objects are equal
+     * */
+    virtual bool __eq__(const Basic &o) const;
+    //! Structural equality comparator
+    virtual int compare(const Basic &o) const;
+    //! \return `true` if canonical
+    virtual bool is_canonical(const RCP<const Basic> &lhs,
+                              const RCP<const Basic> &rhs) const;
+};
+
+class Equality : public Relational
+{
+    //! Class for operator `==`.
+public:
+    IMPLEMENT_TYPEID(EQUALITY);
+    Equality(const RCP<const Basic> &lhs, const RCP<const Basic> &rhs);
+    virtual RCP<const Basic> create(const RCP<const Basic> &lhs,
+                                    const RCP<const Basic> &rhs) const;
+};
+
+class Unequality : public Relational
+{
+    //! Class for operator `!=`.
+public:
+    IMPLEMENT_TYPEID(UNEQUALITY);
+    Unequality(const RCP<const Basic> &lhs, const RCP<const Basic> &rhs);
+    virtual RCP<const Basic> create(const RCP<const Basic> &lhs,
+                                    const RCP<const Basic> &rhs) const;
+};
+
+class _Less : public Relational
+{
+    //! Not intended for general use.
+    //! `_Less` is only used so that `LessThan` and `StrictLessThan` may
+    //! subclass it for `get_gts()` and `get_lts()` functions
+public:
+    //! Constructor for passing to Relational
+    _Less(const RCP<const Basic> &lhs, const RCP<const Basic> &rhs);
+    //! \return the greater side of the relation
+    RCP<const Basic> get_gts() const;
+    //! \return the lesser side of the relation
+    RCP<const Basic> get_lts() const;
+};
+
+class LessThan : public _Less
+{
+    //! Class for operator `<=`.
+public:
+    IMPLEMENT_TYPEID(LESSTHAN);
+    LessThan(const RCP<const Basic> &lhs, const RCP<const Basic> &rhs);
+    virtual RCP<const Basic> create(const RCP<const Basic> &lhs,
+                                    const RCP<const Basic> &rhs) const;
+};
+
+class StrictLessThan : public _Less
+{
+    //! Class for operator `<`.
+public:
+    IMPLEMENT_TYPEID(STRICTLESSTHAN);
+    StrictLessThan(const RCP<const Basic> &lhs, const RCP<const Basic> &rhs);
+    virtual RCP<const Basic> create(const RCP<const Basic> &lhs,
+                                    const RCP<const Basic> &rhs) const;
+};
+
+inline bool is_a_Relational(const Basic &b)
+{
+    return (b.get_type_code() == EQUALITY || b.get_type_code() == UNEQUALITY
+            || b.get_type_code() == LESSTHAN
+            || b.get_type_code() == STRICTLESSTHAN);
+}
+
+//! Returns the canonicalized Equality object from the two arguments
+RCP<const Basic> Eq(const RCP<const Basic> &lhs);
```

Comment:
Its a convenience function, when only a single argument is provided, automatically taking the `rhs` to be `zero`.